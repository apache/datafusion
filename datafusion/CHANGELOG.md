<!---
  Licensed to the Apache Software Foundation (ASF) under one
  or more contributor license agreements.  See the NOTICE file
  distributed with this work for additional information
  regarding copyright ownership.  The ASF licenses this file
  to you under the Apache License, Version 2.0 (the
  "License"); you may not use this file except in compliance
  with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing,
  software distributed under the License is distributed on an
  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  KIND, either express or implied.  See the License for the
  specific language governing permissions and limitations
  under the License.
-->

# Changelog

- [27.0.0](../dev/changelog/27.0.0.md)
- [26.0.0](../dev/changelog/26.0.0.md)
- [25.0.0](../dev/changelog/25.0.0.md)
- [24.0.0](../dev/changelog/24.0.0.md)
- [23.0.0](../dev/changelog/23.0.0.md)
- [22.0.0](../dev/changelog/22.0.0.md)
- [21.1.0](../dev/changelog/21.1.0.md)
- [21.0.0](../dev/changelog/21.0.0.md)
- [20.0.0](../dev/changelog/20.0.0.md)
- [19.0.0](../dev/changelog/19.0.0.md)
- [18.0.0](../dev/changelog/18.0.0.md)
- [17.0.0](../dev/changelog/17.0.0.md)
- [16.1.0](../dev/changelog/16.1.0.md)
- [16.0.0](../dev/changelog/16.0.0.md)
- [15.0.0](../dev/changelog/15.0.0.md)
- [14.0.0](../dev/changelog/14.0.0.md)
- [13.0.0](../dev/changelog/13.0.0.md)
- [12.0.0](../dev/changelog/12.0.0.md)
- [11.0.0](../dev/changelog/11.0.0.md)
- [10.0.0](../dev/changelog/10.0.0.md)
- [9.0.0](../dev/changelog/9.0.0.md)
- [8.0.0](../dev/changelog/8.0.0.md)
- [7.1.0](../dev/changelog/7.1.0.md)
- [7.0.0](../dev/changelog/7.0.0.md)
- [6.0.0](../dev/changelog/6.0.0.md)
- [5.0.0](../dev/changelog/5.0.0.md)
