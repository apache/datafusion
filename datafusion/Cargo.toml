# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

[package]
name = "datafusion"
description = "DataFusion is an in-memory query engine that uses Apache Arrow as the memory model"
version = "6.0.0"
homepage = "https://github.com/apache/arrow-datafusion"
repository = "https://github.com/apache/arrow-datafusion"
readme = "../README.md"
authors = ["Apache Arrow <dev@arrow.apache.org>"]
license = "Apache-2.0"
keywords = [ "arrow", "query", "sql" ]
include = [
    "benches/*.rs",
    "src/**/*.rs",
    "Cargo.toml",
]
edition = "2021"
rust-version = "1.58"

[lib]
name = "datafusion"
path = "src/lib.rs"

[features]
default = ["crypto_expressions", "regex_expressions", "unicode_expressions"]
simd = ["arrow/simd"]
crypto_expressions = ["md-5", "sha2", "blake2", "blake3"]
regex_expressions = ["regex"]
unicode_expressions = ["unicode-segmentation"]
pyarrow = ["pyo3", "arrow/pyarrow", "datafusion-common/pyarrow"]
# Used for testing ONLY: causes all values to hash to the same value (test for collisions)
force_hash_collisions = []
# Used to enable the avro format
avro = ["avro-rs", "num-traits", "datafusion-common/avro"]

[dependencies]
datafusion-common = { path = "../datafusion-common", version = "6.0.0" }
ahash = { version = "0.7", default-features = false }
hashbrown = { version = "0.12", features = ["raw"] }
arrow = { version = "8.0.0", features = ["prettyprint"] }
parquet = { version = "8.0.0", features = ["arrow"] }
sqlparser = "0.13"
paste = "^1.0"
num_cpus = "1.13.0"
chrono = { version = "0.4", default-features = false }
async-trait = "0.1.41"
futures = "0.3"
pin-project-lite= "^0.2.7"
tokio = { version = "1.0", features = ["macros", "rt", "rt-multi-thread", "sync", "fs", "parking_lot"] }
tokio-stream = "0.1"
log = "^0.4"
md-5 = { version = "^0.10.0", optional = true }
sha2 = { version = "^0.10.1", optional = true }
blake2 = { version = "^0.10.2", optional = true }
blake3 = { version = "1.0", optional = true }
ordered-float = "2.10"
unicode-segmentation = { version = "^1.7.1", optional = true }
regex = { version = "^1.4.3", optional = true }
lazy_static = { version = "^1.4.0" }
smallvec = { version = "1.6", features = ["union"] }
rand = "0.8"
avro-rs = { version = "0.13", features = ["snappy"], optional = true }
num-traits = { version = "0.2", optional = true }
pyo3 = { version = "0.15", optional = true }
tempfile = "3"
parking_lot = "0.12"

[dev-dependencies]
criterion = "0.3"
doc-comment = "0.3"
fuzz-utils = { path = "fuzz-utils" }

[[bench]]
name = "aggregate_query_sql"
harness = false

[[bench]]
name = "sort_limit_query_sql"
harness = false

[[bench]]
name = "math_query_sql"
harness = false

[[bench]]
name = "filter_query_sql"
harness = false

[[bench]]
name = "window_query_sql"
harness = false

[[bench]]
name = "scalar"
harness = false

[[bench]]
name = "physical_plan"
harness = false
