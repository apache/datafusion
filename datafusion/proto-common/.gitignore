# Files generated by regen.sh
proto/proto_descriptor.bin
src/datafusion_common.rs
src/datafusion_common.serde.rs
