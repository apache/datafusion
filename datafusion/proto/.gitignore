# Files generated by regen.sh
proto/proto_descriptor.bin
src/datafusion.rs
datafusion.serde.rs
