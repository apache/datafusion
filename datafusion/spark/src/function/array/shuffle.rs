// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use crate::function::functions_nested_utils::make_scalar_function;
use arrow::array::{
    Array, ArrayRef, Capacities, FixedSizeListArray, GenericListArray, MutableArrayData,
    OffsetSizeTrait,
};
use arrow::buffer::OffsetBuffer;
use arrow::datatypes::DataType::{FixedSizeList, LargeList, List, Null};
use arrow::datatypes::{DataType, FieldRef};
use datafusion_common::cast::{
    as_fixed_size_list_array, as_large_list_array, as_list_array,
};
use datafusion_common::{exec_err, utils::take_function_args, Result};
use datafusion_expr::{ColumnarValue, ScalarUDFImpl, Signature, Volatility};
use rand::rng;
use rand::seq::SliceRandom;
use std::any::Any;
use std::sync::Arc;

#[derive(Debug, PartialEq, Eq, Hash)]
pub struct SparkShuffle {
    signature: Signature,
}

impl Default for SparkShuffle {
    fn default() -> Self {
        Self::new()
    }
}

impl SparkShuffle {
    pub fn new() -> Self {
        Self {
            signature: Signature::arrays(1, None, Volatility::Volatile),
        }
    }
}

impl ScalarUDFImpl for SparkShuffle {
    fn as_any(&self) -> &dyn Any {
        self
    }

    fn name(&self) -> &str {
        "shuffle"
    }

    fn signature(&self) -> &Signature {
        &self.signature
    }

    fn return_type(&self, arg_types: &[DataType]) -> Result<DataType> {
        Ok(arg_types[0].clone())
    }

    fn invoke_with_args(
        &self,
        args: datafusion_expr::ScalarFunctionArgs,
    ) -> Result<ColumnarValue> {
        make_scalar_function(array_shuffle_inner)(&args.args)
    }
}

/// array_shuffle SQL function
pub fn array_shuffle_inner(arg: &[ArrayRef]) -> Result<ArrayRef> {
    let [input_array] = take_function_args("shuffle", arg)?;
    match &input_array.data_type() {
        List(field) => {
            let array = as_list_array(input_array)?;
            general_array_shuffle::<i32>(array, field)
        }
        LargeList(field) => {
            let array = as_large_list_array(input_array)?;
            general_array_shuffle::<i64>(array, field)
        }
        FixedSizeList(field, _) => {
            let array = as_fixed_size_list_array(input_array)?;
            fixed_size_array_shuffle(array, field)
        }
        Null => Ok(Arc::clone(input_array)),
        array_type => exec_err!("shuffle does not support type '{array_type}'."),
    }
}

fn general_array_shuffle<O: OffsetSizeTrait>(
    array: &GenericListArray<O>,
    field: &FieldRef,
) -> Result<ArrayRef> {
    let values = array.values();
    let original_data = values.to_data();
    let capacity = Capacities::Array(original_data.len());
    let mut offsets = vec![O::usize_as(0)];
    let mut nulls = vec![];
    let mut mutable =
        MutableArrayData::with_capacities(vec![&original_data], false, capacity);
    let mut rng = rng();

    for (row_index, offset_window) in array.offsets().windows(2).enumerate() {
        // skip the null value
        if array.is_null(row_index) {
            nulls.push(false);
            offsets.push(offsets[row_index] + O::one());
            mutable.extend(0, 0, 1);
            continue;
        }
        nulls.push(true);
        let start = offset_window[0];
        let end = offset_window[1];
        let length = (end - start).to_usize().unwrap();

        // Create indices and shuffle them
        let mut indices: Vec<usize> =
            (start.to_usize().unwrap()..end.to_usize().unwrap()).collect();
        indices.shuffle(&mut rng);

        // Add shuffled elements
        for &index in &indices {
            mutable.extend(0, index, index + 1);
        }

        offsets.push(offsets[row_index] + O::usize_as(length));
    }

    let data = mutable.freeze();
    Ok(Arc::new(GenericListArray::<O>::try_new(
        Arc::clone(field),
        OffsetBuffer::<O>::new(offsets.into()),
        arrow::array::make_array(data),
        Some(nulls.into()),
    )?))
}

fn fixed_size_array_shuffle(
    array: &FixedSizeListArray,
    field: &FieldRef,
) -> Result<ArrayRef> {
    let values = array.values();
    let original_data = values.to_data();
    let capacity = Capacities::Array(original_data.len());
    let mut nulls = vec![];
    let mut mutable =
        MutableArrayData::with_capacities(vec![&original_data], false, capacity);
    let value_length = array.value_length() as usize;
    let mut rng = rng();

    for row_index in 0..array.len() {
        // skip the null value
        if array.is_null(row_index) {
            nulls.push(false);
            mutable.extend(0, 0, value_length);
            continue;
        }
        nulls.push(true);

        let start = row_index * value_length;
        let end = start + value_length;

        // Create indices and shuffle them
        let mut indices: Vec<usize> = (start..end).collect();
        indices.shuffle(&mut rng);

        // Add shuffled elements
        for &index in &indices {
            mutable.extend(0, index, index + 1);
        }
    }

    let data = mutable.freeze();
    Ok(Arc::new(FixedSizeListArray::try_new(
        Arc::clone(field),
        array.value_length(),
        arrow::array::make_array(data),
        Some(nulls.into()),
    )?))
}
