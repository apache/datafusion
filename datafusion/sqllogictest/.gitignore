*.py
test_files/tpch/data