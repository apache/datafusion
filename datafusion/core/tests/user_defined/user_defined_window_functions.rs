// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

//! This module contains end to end tests of creating
//! user defined window functions

use std::{
    ops::Range,
    sync::{
        atomic::{AtomicUsize, Ordering},
        Arc,
    },
};

use arrow::array::AsArray;
use arrow_array::{ArrayRef, Int64Array, RecordBatch, StringArray};
use arrow_schema::DataType;
use datafusion::{assert_batches_eq, prelude::SessionContext};
use datafusion_common::{Result, ScalarValue};
use datafusion_expr::{
    function::PartitionEvaluatorFactory, window_state::WindowAggState,
    PartitionEvaluator, ReturnTypeFunction, Signature, Volatility, WindowUDF,
};

/// A query with a window function evaluated over the entire partition
const UNBOUNDED_WINDOW_QUERY: &str = "SELECT x, y, val, \
     odd_counter(val) OVER (PARTITION BY x ORDER BY y) \
     from t ORDER BY x, y";

/// A query with a window function evaluated over a moving window
const BOUNDED_WINDOW_QUERY:  &str  =
    "SELECT x, y, val, \
     odd_counter(val) OVER (PARTITION BY x ORDER BY y ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING) \
     from t ORDER BY x, y";

/// Test to show the contents of the setup
#[tokio::test]
async fn test_setup() {
    let test_state = TestState::new();
    let TestContext { ctx, test_state: _ } = TestContext::new(test_state);

    let sql = "SELECT * from t order by x, y";
    let expected = vec![
        "+---+---+-----+",
        "| x | y | val |",
        "+---+---+-----+",
        "| 1 | a | 0   |",
        "| 1 | b | 1   |",
        "| 1 | c | 2   |",
        "| 2 | d | 3   |",
        "| 2 | e | 4   |",
        "| 2 | f | 5   |",
        "| 2 | g | 6   |",
        "| 2 | h | 6   |",
        "| 2 | i | 6   |",
        "| 2 | j | 6   |",
        "+---+---+-----+",
    ];
    assert_batches_eq!(expected, &execute(&ctx, sql).await.unwrap());
}

/// Basic user defined window function
#[tokio::test]
async fn test_udwf() {
    let test_state = TestState::new();
    let TestContext { ctx, test_state } = TestContext::new(test_state);

    let expected = vec![
        "+---+---+-----+--------------------+",
        "| x | y | val | odd_counter(t.val) |",
        "+---+---+-----+--------------------+",
        "| 1 | a | 0   | 1                  |",
        "| 1 | b | 1   | 1                  |",
        "| 1 | c | 2   | 1                  |",
        "| 2 | d | 3   | 2                  |",
        "| 2 | e | 4   | 2                  |",
        "| 2 | f | 5   | 2                  |",
        "| 2 | g | 6   | 2                  |",
        "| 2 | h | 6   | 2                  |",
        "| 2 | i | 6   | 2                  |",
        "| 2 | j | 6   | 2                  |",
        "+---+---+-----+--------------------+",
    ];
    assert_batches_eq!(
        expected,
        &execute(&ctx, UNBOUNDED_WINDOW_QUERY).await.unwrap()
    );
    // evaluated on two distinct batches
    assert_eq!(test_state.evaluate_all_called(), 2);
}

/// Basic user defined window function with bounded window
#[tokio::test]
async fn test_udwf_bounded_window_ignores_frame() {
    let test_state = TestState::new();
    let TestContext { ctx, test_state } = TestContext::new(test_state);

    // Since the UDWF doesn't say it needs the window frame, the frame is ignored
    let expected = vec![
        "+---+---+-----+--------------------+",
        "| x | y | val | odd_counter(t.val) |",
        "+---+---+-----+--------------------+",
        "| 1 | a | 0   | 1                  |",
        "| 1 | b | 1   | 1                  |",
        "| 1 | c | 2   | 1                  |",
        "| 2 | d | 3   | 2                  |",
        "| 2 | e | 4   | 2                  |",
        "| 2 | f | 5   | 2                  |",
        "| 2 | g | 6   | 2                  |",
        "| 2 | h | 6   | 2                  |",
        "| 2 | i | 6   | 2                  |",
        "| 2 | j | 6   | 2                  |",
        "+---+---+-----+--------------------+",
    ];
    assert_batches_eq!(
        expected,
        &execute(&ctx, BOUNDED_WINDOW_QUERY).await.unwrap()
    );
    // evaluated on 2 distinct batches (when x=1 and x=2)
    assert_eq!(test_state.evaluate_called(), 0);
    assert_eq!(test_state.evaluate_all_called(), 2);
}

/// Basic user defined window function with bounded window
#[tokio::test]
async fn test_udwf_bounded_window() {
    let test_state = TestState::new().with_uses_window_frame();
    let TestContext { ctx, test_state } = TestContext::new(test_state);

    let expected = vec![
        "+---+---+-----+--------------------+",
        "| x | y | val | odd_counter(t.val) |",
        "+---+---+-----+--------------------+",
        "| 1 | a | 0   | 1                  |",
        "| 1 | b | 1   | 1                  |",
        "| 1 | c | 2   | 1                  |",
        "| 2 | d | 3   | 1                  |",
        "| 2 | e | 4   | 2                  |",
        "| 2 | f | 5   | 1                  |",
        "| 2 | g | 6   | 1                  |",
        "| 2 | h | 6   | 0                  |",
        "| 2 | i | 6   | 0                  |",
        "| 2 | j | 6   | 0                  |",
        "+---+---+-----+--------------------+",
    ];
    assert_batches_eq!(
        expected,
        &execute(&ctx, BOUNDED_WINDOW_QUERY).await.unwrap()
    );
    // Evaluate is called for each input rows
    assert_eq!(test_state.evaluate_called(), 10);
    assert_eq!(test_state.evaluate_all_called(), 0);
}

/// Basic stateful user defined window function
#[tokio::test]
async fn test_stateful_udwf() {
    let test_state = TestState::new()
        .with_supports_bounded_execution()
        .with_uses_window_frame();
    let TestContext { ctx, test_state } = TestContext::new(test_state);

    let expected = vec![
        "+---+---+-----+--------------------+",
        "| x | y | val | odd_counter(t.val) |",
        "+---+---+-----+--------------------+",
        "| 1 | a | 0   | 0                  |",
        "| 1 | b | 1   | 1                  |",
        "| 1 | c | 2   | 1                  |",
        "| 2 | d | 3   | 1                  |",
        "| 2 | e | 4   | 1                  |",
        "| 2 | f | 5   | 2                  |",
        "| 2 | g | 6   | 2                  |",
        "| 2 | h | 6   | 2                  |",
        "| 2 | i | 6   | 2                  |",
        "| 2 | j | 6   | 2                  |",
        "+---+---+-----+--------------------+",
    ];
    assert_batches_eq!(
        expected,
        &execute(&ctx, UNBOUNDED_WINDOW_QUERY).await.unwrap()
    );
    assert_eq!(test_state.evaluate_called(), 10);
    assert_eq!(test_state.update_state_called(), 10);
    assert_eq!(test_state.evaluate_all_called(), 0);
}

/// Basic stateful user defined window function with bounded window
#[tokio::test]
async fn test_stateful_udwf_bounded_window() {
    let test_state = TestState::new()
        .with_supports_bounded_execution()
        .with_uses_window_frame();
    let TestContext { ctx, test_state } = TestContext::new(test_state);

    let expected = vec![
        "+---+---+-----+--------------------+",
        "| x | y | val | odd_counter(t.val) |",
        "+---+---+-----+--------------------+",
        "| 1 | a | 0   | 1                  |",
        "| 1 | b | 1   | 1                  |",
        "| 1 | c | 2   | 1                  |",
        "| 2 | d | 3   | 1                  |",
        "| 2 | e | 4   | 2                  |",
        "| 2 | f | 5   | 1                  |",
        "| 2 | g | 6   | 1                  |",
        "| 2 | h | 6   | 0                  |",
        "| 2 | i | 6   | 0                  |",
        "| 2 | j | 6   | 0                  |",
        "+---+---+-----+--------------------+",
    ];
    assert_batches_eq!(
        expected,
        &execute(&ctx, BOUNDED_WINDOW_QUERY).await.unwrap()
    );
    // Evaluate and update_state is called for each input row
    assert_eq!(test_state.evaluate_called(), 10);
    assert_eq!(test_state.update_state_called(), 10);
    assert_eq!(test_state.evaluate_all_called(), 0);
}

/// user defined window function using rank
#[tokio::test]
async fn test_udwf_query_include_rank() {
    let test_state = TestState::new().with_include_rank();
    let TestContext { ctx, test_state } = TestContext::new(test_state);

    let expected = vec![
        "+---+---+-----+--------------------+",
        "| x | y | val | odd_counter(t.val) |",
        "+---+---+-----+--------------------+",
        "| 1 | a | 0   | 3                  |",
        "| 1 | b | 1   | 2                  |",
        "| 1 | c | 2   | 1                  |",
        "| 2 | d | 3   | 7                  |",
        "| 2 | e | 4   | 6                  |",
        "| 2 | f | 5   | 5                  |",
        "| 2 | g | 6   | 4                  |",
        "| 2 | h | 6   | 3                  |",
        "| 2 | i | 6   | 2                  |",
        "| 2 | j | 6   | 1                  |",
        "+---+---+-----+--------------------+",
    ];
    assert_batches_eq!(
        expected,
        &execute(&ctx, UNBOUNDED_WINDOW_QUERY).await.unwrap()
    );
    assert_eq!(test_state.evaluate_called(), 0);
    assert_eq!(test_state.evaluate_all_called(), 0);
    // evaluated on 2 distinct batches (when x=1 and x=2)
    assert_eq!(test_state.evaluate_all_with_rank_called(), 2);
}

/// user defined window function with bounded window using rank
#[tokio::test]
async fn test_udwf_bounded_query_include_rank() {
    let test_state = TestState::new().with_include_rank();
    let TestContext { ctx, test_state } = TestContext::new(test_state);

    let expected = vec![
        "+---+---+-----+--------------------+",
        "| x | y | val | odd_counter(t.val) |",
        "+---+---+-----+--------------------+",
        "| 1 | a | 0   | 3                  |",
        "| 1 | b | 1   | 2                  |",
        "| 1 | c | 2   | 1                  |",
        "| 2 | d | 3   | 7                  |",
        "| 2 | e | 4   | 6                  |",
        "| 2 | f | 5   | 5                  |",
        "| 2 | g | 6   | 4                  |",
        "| 2 | h | 6   | 3                  |",
        "| 2 | i | 6   | 2                  |",
        "| 2 | j | 6   | 1                  |",
        "+---+---+-----+--------------------+",
    ];
    assert_batches_eq!(
        expected,
        &execute(&ctx, BOUNDED_WINDOW_QUERY).await.unwrap()
    );
    assert_eq!(test_state.evaluate_called(), 0);
    assert_eq!(test_state.evaluate_all_called(), 0);
    // evaluated on 2 distinct batches (when x=1 and x=2)
    assert_eq!(test_state.evaluate_all_with_rank_called(), 2);
}

async fn execute(ctx: &SessionContext, sql: &str) -> Result<Vec<RecordBatch>> {
    ctx.sql(sql).await?.collect().await
}

/// Returns an context with a table "t" and the "first" and "time_sum"
/// aggregate functions registered.
///
/// "t" contains this data:
///
/// ```text
/// x | y | val
/// 1 | a | 0
/// 1 | b | 1
/// 1 | c | 2
/// 2 | d | 3
/// 2 | e | 4
/// 2 | f | 5
/// 2 | g | 6
/// 2 | h | 6
/// 2 | i | 6
/// 2 | j | 6
/// ```
struct TestContext {
    ctx: SessionContext,
    test_state: Arc<TestState>,
}

impl TestContext {
    fn new(test_state: TestState) -> Self {
        let test_state = Arc::new(test_state);
        let x = Int64Array::from(vec![1, 1, 1, 2, 2, 2, 2, 2, 2, 2]);
        let y = StringArray::from(vec!["a", "b", "c", "d", "e", "f", "g", "h", "i", "j"]);
        let val = Int64Array::from(vec![0, 1, 2, 3, 4, 5, 6, 6, 6, 6]);

        let batch = RecordBatch::try_from_iter(vec![
            ("x", Arc::new(x) as _),
            ("y", Arc::new(y) as _),
            ("val", Arc::new(val) as _),
        ])
        .unwrap();

        let mut ctx = SessionContext::new();

        ctx.register_batch("t", batch).unwrap();

        // Tell DataFusion about the window function
        OddCounter::register(&mut ctx, Arc::clone(&test_state));

        Self { ctx, test_state }
    }
}

#[derive(Debug, Default)]
struct TestState {
    /// How many times was `evaluate_all` called?
    evaluate_all_called: AtomicUsize,
    /// How many times was `evaluate` called?
    evaluate_called: AtomicUsize,
    /// How many times was `update_state` called?
    update_state_called: AtomicUsize,
    /// How many times was `evaluate_all_with_rank` called?
    evaluate_all_with_rank_called: AtomicUsize,
    /// should the functions say they use the window frame?
    uses_window_frame: bool,
    /// should the functions say they support bounded execution
    supports_bounded_execution: bool,
    /// should the functions they need include rank
    include_rank: bool,
}

impl TestState {
    fn new() -> Self {
        Default::default()
    }

    /// Set that this function should use the window frame
    fn with_uses_window_frame(mut self) -> Self {
        self.uses_window_frame = true;
        self
    }

    /// Set that this function should use bounded / stateful execution
    fn with_supports_bounded_execution(mut self) -> Self {
        self.supports_bounded_execution = true;
        self
    }

    /// Set that this function should include rank
    fn with_include_rank(mut self) -> Self {
        self.include_rank = true;
        self
    }

    /// return the evaluate_all_called counter
    fn evaluate_all_called(&self) -> usize {
        self.evaluate_all_called.load(Ordering::SeqCst)
    }

    /// update the evaluate_all_called counter
    fn inc_evaluate_all_called(&self) {
        self.evaluate_all_called.fetch_add(1, Ordering::SeqCst);
    }

    /// return the evaluate_called counter
    fn evaluate_called(&self) -> usize {
        self.evaluate_called.load(Ordering::SeqCst)
    }

    /// update the evaluate_called counter
    fn inc_evaluate_called(&self) {
        self.evaluate_called.fetch_add(1, Ordering::SeqCst);
    }

    /// return the update_state_called counter
    fn update_state_called(&self) -> usize {
        self.update_state_called.load(Ordering::SeqCst)
    }

    /// update the update_state_called counter
    fn inc_update_state_called(&self) {
        self.update_state_called.fetch_add(1, Ordering::SeqCst);
    }

    /// return the evaluate_all_with_rank_called counter
    fn evaluate_all_with_rank_called(&self) -> usize {
        self.evaluate_all_with_rank_called.load(Ordering::SeqCst)
    }

    /// update the evaluate_all_with_rank_called counter
    fn inc_evaluate_all_with_rank_called(&self) {
        self.evaluate_all_with_rank_called
            .fetch_add(1, Ordering::SeqCst);
    }
}

// Partition Evaluator that counts the number of odd numbers in the window frame using evaluate
#[derive(Debug)]
struct OddCounter {
    test_state: Arc<TestState>,
}

impl OddCounter {
    fn new(test_state: Arc<TestState>) -> Self {
        Self { test_state }
    }

    fn register(ctx: &mut SessionContext, test_state: Arc<TestState>) {
        let name = "odd_counter";
        let volatility = Volatility::Immutable;

        let signature = Signature::exact(vec![DataType::Int64], volatility);

        let return_type = Arc::new(DataType::Int64);
        let return_type: ReturnTypeFunction =
            Arc::new(move |_| Ok(Arc::clone(&return_type)));

        let partition_evaluator_factory: PartitionEvaluatorFactory =
            Arc::new(move || Ok(Box::new(OddCounter::new(Arc::clone(&test_state)))));

        ctx.register_udwf(WindowUDF::new(
            name,
            &signature,
            &return_type,
            &partition_evaluator_factory,
        ))
    }
}

impl PartitionEvaluator for OddCounter {
    fn evaluate(
        &mut self,
        values: &[ArrayRef],
        range: &Range<usize>,
    ) -> Result<ScalarValue> {
        println!("evaluate, values: {values:#?}, range: {range:?}");

        self.test_state.inc_evaluate_called();
        let values: &Int64Array = values.get(0).unwrap().as_primitive();
        let values = values.slice(range.start, range.len());
        let scalar = ScalarValue::Int64(Some(odd_count(&values)));
        Ok(scalar)
    }

    fn evaluate_all(
        &mut self,
        values: &[arrow_array::ArrayRef],
        num_rows: usize,
    ) -> Result<arrow_array::ArrayRef> {
        println!("evaluate_all, values: {values:#?}, num_rows: {num_rows}");

        self.test_state.inc_evaluate_all_called();
        Ok(odd_count_arr(
            values.get(0).unwrap().as_primitive(),
            num_rows,
        ))
    }

    fn evaluate_all_with_rank(
        &self,
        num_rows: usize,
        ranks_in_partition: &[Range<usize>],
    ) -> Result<ArrayRef> {
        self.test_state.inc_evaluate_all_with_rank_called();
        println!("evaluate_all_with_rank, values: {num_rows:#?}, ranks_in_partitions: {ranks_in_partition:?}");
        // when evaluating with ranks, just return the inverse rank instead
        let array: Int64Array = ranks_in_partition
            .iter()
            // cloned range is an iterator
            .cloned()
            .flatten()
            .map(|v| (num_rows - v) as i64)
            .collect();
        Ok(Arc::new(array))
    }

    fn update_state(
        &mut self,
        _state: &WindowAggState,
        _idx: usize,
        _range_columns: &[ArrayRef],
        _sort_partition_points: &[Range<usize>],
    ) -> Result<()> {
        self.test_state.inc_update_state_called();
        Ok(())
    }

    fn supports_bounded_execution(&self) -> bool {
        self.test_state.supports_bounded_execution
    }

    fn uses_window_frame(&self) -> bool {
        self.test_state.uses_window_frame
    }

    fn include_rank(&self) -> bool {
        self.test_state.include_rank
    }
}

/// returns the number of entries in arr that are odd
fn odd_count(arr: &Int64Array) -> i64 {
    arr.iter().filter_map(|x| x.map(|x| x % 2)).sum()
}

/// returns an array of num_rows that has the number of odd values in `arr`
fn odd_count_arr(arr: &Int64Array, num_rows: usize) -> arrow_array::ArrayRef {
    let array: Int64Array = std::iter::repeat(odd_count(arr)).take(num_rows).collect();
    Arc::new(array)
}
