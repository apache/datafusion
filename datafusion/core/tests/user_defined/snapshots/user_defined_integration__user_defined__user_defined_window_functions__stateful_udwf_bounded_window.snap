---
source: datafusion/core/tests/user_defined/user_defined_window_functions.rs
expression: fmt_batches(&actual)
---
+---+---+-----+--------------------------------------------------------------------------------------------------------------+
| x | y | val | odd_counter(t.val) PARTITION BY [t.x] ORDER BY [t.y ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING |
+---+---+-----+--------------------------------------------------------------------------------------------------------------+
| 1 | a | 0   | 1                                                                                                            |
| 1 | b | 1   | 1                                                                                                            |
| 1 | c | 2   | 1                                                                                                            |
| 2 | d | 3   | 1                                                                                                            |
| 2 | e | 4   | 2                                                                                                            |
| 2 | f | 5   | 1                                                                                                            |
| 2 | g | 6   | 1                                                                                                            |
| 2 | h | 6   | 0                                                                                                            |
| 2 | i | 6   | 0                                                                                                            |
| 2 | j | 6   | 0                                                                                                            |
+---+---+-----+--------------------------------------------------------------------------------------------------------------+
